src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs

```diff
@@ -64,6 +66,588 @@ public static void ToStringTest()
+        }
+
+        [Fact]
+        public void NoSuchPropertyExplicitlyNoIndices()
+        {
+            ConstantExpression instance = Expression.Constant("");
+            Assert.Throws<ArgumentException>(() => Expression.Property(instance, "ThisDoesNotExist", Array.Empty<Expression>()));
+            Assert.Throws<ArgumentException>(() => Expression.Property(instance, "ThisDoesNotExist", null));
+        }
+
+        [Theory, ClassData(typeof(CompilationTypes))]
+        public void NonIndexedPropertyExplicitlyNoIndices(bool useInterpreter)
+        {
+            ConstantExpression instance = Expression.Constant("123");
+            IndexExpression prop = Expression.Property(instance, "Length", null);
+            Expression<Func<int>> exp = Expression.Lambda<Func<int>>(prop);
+            Func<int> func = exp.Compile(useInterpreter);
+            Assert.Equal(3, func());
+        }
+
+        [Fact]
+        public void FindNothingForNullArgument()
+        {
+            ConstantExpression instance = Expression.Constant("123");
+            Assert.Throws<ArgumentException>(() => Expression.Property(instance, "Length", new Expression[] {null}));
+        }
+
+        [Fact]
+        public void NullArgument()
+        {
+            ConstantExpression instance = Expression.Constant(new Dictionary<int, int>());
+            PropertyInfo prop = typeof(Dictionary<int, int>).GetProperty("Item");
+            Assert.Throws<ArgumentException>(() => Expression.Property(instance, "Item", new Expression[] {null}));
+            Assert.Throws<ArgumentNullException>("arguments[0]", () => Expression.Property(instance, prop, new Expression[] {null}));
+        }
+
+        [Fact]
+        public void UnreadableIndex()
+        {
+            ConstantExpression instance = Expression.Constant(new Dictionary<int, int>());
+            PropertyInfo prop = typeof(Dictionary<int, int>).GetProperty("Item");
+            MemberExpression index = Expression.Property(null, typeof(Unreadable<int>).GetProperty(nameof(Unreadable<int>.WriteOnly)));
+            Assert.Throws<ArgumentException>("arguments[0]", () => Expression.Property(instance, "Item", index));
+            Assert.Throws<ArgumentException>("arguments[0]", () => Expression.Property(instance, prop, index));
+        }
+
+        private class IntAndExpressionIndexed
+        {
+            public bool this[int x, Expression<Action> y] => true;
+        }
+
+        private class OverloadedIndexers
+        {
+            public int this[int index] => 0;
+
+            public int this[int x, int y] => 1;
+
+            public int this[string index] => 2;
+        }
+
+        private class OverloadedIndexersBothMatchString
+        {
+            public int this[IComparable index] => 0;
+
+            public int this[ICloneable index] => 1;
+        }
+
+        private static class Unreadable<T>
```

Comment:
We have an identical class defined here: https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/tests/HelperTypes.cs#L271